File: easy/3823_reverse_letters_then_special_characters_in_a_string/solution.py
class Solution:
    def reverseByType(self, s: str) -> str:
        chars = list(s)

        for should_be_lower_case in (True, False):
            left = 0
            right = len(s) - 1

            while True:
                while (
                    left < right
                    and chars[left].islower() == should_be_lower_case
                ):
                    left += 1
                while (
                    right > 0
                    and chars[right].islower() == should_be_lower_case
                ):
                    right -= 1

                if left >= right:
                    break

                chars[left], chars[right] = chars[right], chars[left]

                left += 1
                right -= 1

        return "".join(chars)

File: easy/3823_reverse_letters_then_special_characters_in_a_string/Solution.java
class Solution {
  public String reverseByType(String s) {
    final char[] chars = s.toCharArray();

    for (final boolean shouldBeLowerCase : new boolean[] {true, false}) {
      int left = 0;
      int right = s.length() - 1;

      while (true) {
        while (left < right && Character.isLowerCase(chars[left]) == shouldBeLowerCase) {
          ++left;
        }

        while (right > 0 && Character.isLowerCase(chars[right]) == shouldBeLowerCase) {
          --right;
        }

        if (left >= right) {
          break;
        }

        final char temp = chars[left];
        chars[left] = chars[right];
        chars[right] = temp;

        ++left;
        --right;
      }
    }

    return String.valueOf(chars);
  }
}
